kombu/transport/azureservicebus.py

```diff
@@ -0,0 +1,154 @@
+from kombu.utils.objects import cached_property
+
+from . import virtual
+
+try:
+    from azure.servicebus import ServiceBusService, Message, Queue
+except ImportError:
+    ServiceBusService = Message = Queue = None
+
+# dots are replaced by dash, all other punctuation replaced by underscore.
+CHARS_REPLACE_TABLE = {
+    ord(c): 0x5f for c in string.punctuation if c not in '_'
+}
+
+
+class Channel(virtual.Channel):
+    """Azure Service Bus channel."""
+
+    default_visibility_timeout = 1800  # 30 minutes.
+    domain_format = 'kombu%(vhost)s'
+    _sbq = None
+    _queue_cache = {}
+
+    def __init__(self, *args, **kwargs):
+        if ServiceBusService is None:
+            raise ImportError('Azure Service Bus transport requires the '
+                              'azure-servicebus library')
+
+        super(Channel, self).__init__(*args, **kwargs)
+
+        for queue in self.sbq.list_queues():
+            self._queue_cache[queue] = queue
+
+    def entity_name(self, name, table=CHARS_REPLACE_TABLE):
+        """Format AMQP queue name into a valid ServiceBus queue name."""
+        return text_t(safe_str(name)).translate(table)
+
+    def _new_queue(self, queue, **kwargs):
+        """Ensure a queue exists in ServiceBus."""
+        queue = self.entity_name(self.queue_name_prefix + queue)
+        try:
+            return self._queue_cache[queue]
+        except KeyError:
+            self.sbq.create_queue(queue, fail_on_exist=False)
+            q = self._queue_cache[queue] = self.sbq.get_queue(queue)
+            return q
+
+    def _delete(self, queue, *args, **kwargs):
+        """Delete queue by name."""
+        queue_name = self.entity_name(queue)
+        self._queue_cache.pop(queue_name, None)
+        self.sbq.delete_queue(queue_name)
+        super(Channel, self)._delete(queue_name)
+
+    def _put(self, queue, message, **kwargs):
+        """Put message onto queue."""
+        msg = Message(dumps(message))
+        self.sbq.send_queue_message(self.entity_name(queue), msg)
+
+    def _get(self, queue, timeout=None):
+        """Try to retrieve a single message off ``queue``."""
+        message = self.sbq.receive_queue_message(self.entity_name(queue),
+                                                 timeout=timeout,
+                                                 peek_lock=False)
+
+        if message.body is None:
+            raise Empty()
+
+        return loads(bytes_to_str(message.body))
+
+    def _size(self, queue):
+        """Return the number of messages in a queue."""
+        return self._new_queue(queue).message_count
+
+    def _purge(self, queue):
+        """Delete all current messages in a queue."""
+        n = 0
+
+        while True:
+            message = self.sbq.read_delete_queue_message(
+                self.entity_name(queue), timeout=0.1)
+
+            if not message.body:
+                break
+            else:
+                n += 1
+
+        return n
+
+    @property
+    def sbq(self):
```

Comment:
Not sure, but perhaps the abbreviation will be difficult to read for readers unfamiliar with the service?

Reply:
Done in a7df447.